### PR TITLE
11.0 imp product_pricelist_direct_print

### PR DESCRIPTION
**Pull request by ernestotejeda**
It must be rebased when PR #370  is merged

* Print price list for future date
* Print price list for N last ordered products
* Send report by email to each partner selected

Cc @Tecnativa

### CODE REVIEW

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+                (6, 0, self.env.context.get('active_ids', []))]
+        elif self.env.context.get('active_model') == 'product.pricelist':
+            res['pricelist_id'] = self.env.context.get('active_id', False)
+        elif self.env.context.get('active_model') == 'res.partner':
+            active_ids = self.env.context.get('active_ids', [])
+            res['partner_ids'] = [(6, 0, active_ids)]
+            if len(active_ids) == 1:
+                res['partner_id'] = active_ids[0]
+                partner = self.env['res.partner'].browse(active_ids[0])
+                res['pricelist_id'] = partner.property_product_pricelist.id
```
Comment (sergio-teruel) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py:82]:
```suggestion
                res['pricelist_id'] = partner.property_product_pricelist.id
        elif self.env.context.get('active_model') == 'product.pricelist.item':
            active_ids = self.env.context.get('active_ids', [])
            items = self.env['product.pricelist.item'].browse(active_ids)
            if items[:1].applied_on == '0_product_variant':
                res['show_variants'] = True
                res['product_ids'] = [(6, 0, items.mapped('product_id').ids)]
            else:
                res['product_tmpl_ids'] = [
                    (6, 0, items.mapped('product_tmpl_id').ids)]               
```

Comment (sergio-teruel) [reply product_pricelist_direct_print/wizards/product_pricelist_print.py:82]: Can you add this code? is for print from pricelist items..
Thanks

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            self.write({
+                'partner_id': partner.id,
+                'pricelist_id': partner.property_product_pricelist.id,
+            })
+            self.force_pricelist_send()
+
+    @api.multi
+    def force_pricelist_send(self):
+        if self.env.context.get('active_model') != 'res.partner':
+            return False
```
Comment (sergio-teruel) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]: Remove this condition to allow send email from any model

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        template_id = self.env.ref(
+            'product_pricelist_direct_print.email_template_edi_pricelist').id
+        composer = self.env['mail.compose.message'].with_context({
+            'default_composition_mode': 'mass_mail',
+            'default_notify': True,
+            'default_res_id': self.id,
+            'default_model': 'product.pricelist.print',
+            'default_template_id': template_id,
+            'active_ids': self.ids,
+            'partner_to': self.partner_id,
```
Comment (sergio-teruel) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]: Remove partner_to key

File: product_pricelist_direct_print/wizards/product_pricelist_print_view.xml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                            <field name="name"/>
+                        </tree>
+                    </field>
+                    <field name="product_ids"
+                           attrs="{'invisible':[('show_variants','=',False)]}">
+                        <tree>
+                            <field name="name"/>
+                        </tree>
+                    </field>
+                </page>
```
Comment (sergio-teruel) [inline product_pricelist_direct_print/wizards/product_pricelist_print_view.xml:45]:
```suggestion
                </page>
                <page string="Customers">
                    <field name="partner_ids"/>
                    <field name="partner_count" invisible="1"/>
                </page>                
```

Comment (sergio-teruel) [reply product_pricelist_direct_print/wizards/product_pricelist_print_view.xml:45]: Allow select partners.

File: product_pricelist_direct_print/wizards/product_pricelist_print_view.xml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    src_model="res.partner"
+    groups="product.group_sale_pricelist"
+    view_mode="form" target="new" view_type="form"/>
+
+<act_window id="action_pricelist_print"
+    name="Print Price List"
+    res_model="product.pricelist.print"
+    groups="product.group_sale_pricelist"
+    view_mode="form" target="new" view_type="form"/>
+
```
Comment (sergio-teruel) [inline product_pricelist_direct_print/wizards/product_pricelist_print_view.xml:101]:
```suggestion

<act_window id="action_item_pricelist_print"
    key2="client_print_multi"
    name="Price List"
    res_model="product.pricelist.print"
    src_model="product.pricelist.item"
    groups="product.group_sale_pricelist"
    view_mode="form" target="new" view_type="form"/>

```

Comment (sergio-teruel) [reply product_pricelist_direct_print/wizards/product_pricelist_print_view.xml:101]: Print from pricelists items

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
ping @ernestotejeda

**Comment by ernestotejeda** [comment]
Thanks @sergio-teruel, I have made almost all the changes you propose except the ones related with pricelist items because I don't know where I can find their corresponding tree view to be able to test those changes.
When you talk about selecting several partners, to be consistent with the behavior of the module should be sent to each one an email with the report corresponding to the associated price list, and therefore I have added some other things like, Once at least one partner is selected, the pricelists in the wizard is not visible, and some details about the visibility of the fields I changed too.

**Comment by pedrobaeza** [comment]
You can rebase this PR now that I have merged the other one.

**Comment by chienandalu** [comment]
Well there's no such pricelist item list view :sweat_smile:

**Review by chienandalu — COMMENTED** [review]
I can't make multiple partner send work on runbot:

image

**Comment by ernestotejeda** [comment]
@chienandalu is true that in *Settings -> Technical -> Email -> Messages* the corresponding messages have the field *Recipients* empty, but emails are sent correctly.
If you test with incorrect addresses you will see in *Settings -> Technical -> Email -> Emails* that the emails have the field *To (Partners) * correctly, but if you test with correct addresses, emails are deleted but those ones are sent correctly .
Apparently this is the behavior of mass mailing. When module *mass_mailing* is installed and you send a newsletter you will see the same behavior.

**Comment by pedrobaeza** [comment]
> Thanks @sergio-teruel, I have made almost all the changes you propose except the ones related with pricelist items because I don't know where I can find their corresponding tree view to be able to test those changes.

@ernestotejeda this is because in standard there's no such view... unless you go to the model in *Settings > Technical > Database structure > Model*, find that model, and add menu item. Please do it and add the suggested code.

**Comment by chienandalu** [comment]
@ernestotejeda I'm trying to PR to your repo this commit improving the pricelist items print but I can't find your fork: https://github.com/Tecnativa/product-attribute/commit/7bbc7a43915c38eafee863fba94d343188948eae

- Sets the pricelist in the wizard if all the items belong to the same one.
- Lets to print items applied on different rules altogether.

ping @sergio-teruel

**Review by chienandalu — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]
I don't like too much to have you fields `partner_id` and `partner_ids` for the same thing, as you can check if `len(partner_ids) == 1`, but for not blocking this more time and not doing diff bigger, let's merge it this way.

**Pull request merged**